### ISSUE TITLE
Setting property to null with composite nullable foreign key incorrectly deletes entire row

### ISSUE THREAD

**Issue by craigbrown** [dotnet/efcore#22196]
In my EF Core code-first model, I have a few entities: `Company`, `Product`, `Variant` and `Subscription`. A `Company` can have a `Subscription` to a `Product`, and additionally that subscription can also specify a particular `Variant` (e.g. Beta), but it doesn't have to - the `Variant` can be null. A `Variant` is specific to a `Product`, so if every product has a Beta variant there would have to be a separate `Variant` row with the value "Beta" for every product. And so I use a composite foreign key between `Subscription` and `Variant` to enforce that `Product`-`Variant` relationship, because if a `Company` has a `Subscription` to a `Product` with no Beta variant, I don't want them to be able to set one.

```csharp
modelBuilder.Entity<Subscription>().HasOne(s => s.Variant).WithMany().HasForeignKey("ProductId", "VariantId").HasPrincipalKey("ProductId", "VariantId");
```

The issue I'm having is that when I have an existing `Subscription` with a non-null `Variant`, and I try to update the `Variant` to be null using EF Core, the entire `Subscription` row is being deleted from the database. I would expect the row to remain, but the value in the `VariantId` column to be set to null.

I _am_ able to accomplish this through SQL, but not EF Core.

I assumed this was just an issue with how I've set up my model, specifically the line of code above since when I remove it the problem goes away, but I can't figure out what the problem is. I was thinking maybe I need to do something extra to let EF Core know that `VariantId` can be null, and in that case to ignore the foreign key, but I think it should be doing that anyway. And it's weird that it would delete the whole row instead of just throwing an error.

### Steps to reproduce

1. Create a code-first model containing three entities - S, P and V, so that P has many V, and S has one of P and one or none of V. Add a foreign key between S and V on the id columns of P and V.

2. Add an instance of each to the database and link them so that the instance of S has one of V.

3. Through EF Core, set the property in the parent to null and call `SaveChangesAsync()` on your context. Rather than just setting the property to null, the entire row for the parent will be deleted.

-OR-

1. Clone [this simple example test repository](https://github.com/craigbrown/TestEfCoreUpdate) and open in Visual Studio.

2. Create a new SQL Server database wherever you like and get the connection string.

3. In Nuget Package Manager, run the following commands:
```
$Env:SqlServerDb = "your connection string goes here"
```
```
Update-Database -StartupProject TestEfCoreUpdate.Migrations
```

4. In the BlazorApp project, open `appsettings.json` and paste in the connection string.

5. Run the BlazorApp project. Change the dropdown to None and click Update. The subscription should still exist in the database but with the Variant set to null. Instead, the subscription is deleted from the database.

### Further technical details

EF Core version: 3.1.7
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.1
Operating system: MacOS 10.15.6
IDE: Visual Studio for Mac 8.7.3

**Comment by ajcvickers** [comment]
@craigbrown In EF Core, if any part of a composite key is null, then the entire key is treated as null. This is because the different null semantics in the database verses .NET can be tricky at the best of times, and when it comes to keys it gets worse because we have to decide if two keys with a null will ever compare as the same--yes in .NET, no in SQL. It would not be impossible to come up with rules as to how this should work, but it hasn't been something we have seen requested before, and the simplicity of the current rule seem to help people in general.

**Comment by craigbrown** [comment]
@ajcvickers thanks for your reply, that makes sense. Any thoughts on why it would be deleting the entire row? If I had a non-composite foreign key that I set to null for a particular row, I don't _think_ that would result in that row being deleted. I believe it would either just set the value in that column to null, or throw an error saying there's no such row in the other table with that key.

**Comment by ajcvickers** [comment]
@craigbrown This could be a "delete orphans" behavior, but that would only kick in by convention if the relationship is required and is therefore configured for cascade deletes. It doesn't seem like this should be the case if part of the FK is nullable--it could be a bug. I'll take a look at your repro code when I get a chance.

**Comment by craigbrown** [comment]
@ajcvickers thank you! No rush, I've worked around it for now by just commenting out that line of code (no new migration was required for it to start working)

**Comment by cesarsouza** [comment]
Hi,

I had opened #23883 thinking EF was having a problem handling FKs where only some properties of the FK are nullable. On further thought, I think that the example I provided there was also running in the issue described above, in which EF is deleting the entity rather than attempting to set it to null.

I've implemented a quick-and-dirty workaround for this in #23883 in case it helps anyone.

**Comment by ferferga** [comment]
Hello

Now that this is fixed [in this commit](https://github.com/dotnet/efcore/commit/cff229a48a9eef6522201018ce0ce3f3f52d5357), any chance this gets backported to EFCore 5 as well? @ajcvickers

Thank you very much in advance.

**Comment by ajcvickers** [comment]
Verified that this is fixed in 6.0 and has same root cause as #23883.

With 5.0:

```
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (11ms) [Parameters=[@p0='1'], CommandType='Text', CommandTimeout='30']
      SET NOCOUNT ON;
      DELETE FROM [Subscription]
      WHERE [SubscriptionId] = @p0;
      SELECT @@ROWCOUNT;
```

With 6.0:

```
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (15ms) [Parameters=[@p1='1', @p0=NULL (DbType = Int32)], CommandType='Text', CommandTimeout='30']
      SET NOCOUNT ON;
      UPDATE [Subscription] SET [VariantId] = @p0
      WHERE [SubscriptionId] = @p1;
      SELECT @@ROWCOUNT;
```

**Comment by ajcvickers** [comment]
@ferferga This issue doesn't [meet the bar](https://docs.microsoft.com/en-us/ef/core/what-is-new/release-planning) for patching in 5.0.
